File: app/src/main/java/com/chooloo/www/callmanager/adapter/RecentsAdapter.java
package com.chooloo.www.callmanager.adapter;

import android.content.Context;
import android.database.Cursor;
import android.provider.ContactsContract;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;

import com.chooloo.www.callmanager.R;
import com.chooloo.www.callmanager.adapter.base.CursorAdapter;
import com.chooloo.www.callmanager.database.entity.RecentCall;
import com.chooloo.www.callmanager.ui2.ListItemHolder;
import com.chooloo.www.callmanager.util.RelativeTime;
import com.chooloo.www.callmanager.util.Utilities;

import org.apache.poi.xdgf.util.Util;

import java.util.Date;

public class RecentsAdapter<VH extends ListItemHolder> extends CursorAdapter<VH> {

    private OnRecentItemClickListener mOnRecentItemClickListener;

    public RecentsAdapter(Context context) {
        super(context);
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return (VH) new ListItemHolder(LayoutInflater.from(mContext).inflate(R.layout.list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, Cursor cursor) {
        RecentCall recentCall = new RecentCall(mContext, cursor);

        String name = recentCall.getCallerName();
        String number = recentCall.getCallerNumber();
        Date date = recentCall.getCallDate();

        holder.setBigText(name == null ? number : name + (recentCall.getCount() > 0 ? " (" + recentCall.getCount() + ")" : ""));
        holder.setSmallText(RelativeTime.getTimeAgo(date.getTime()));
        holder.showPhoto(true, Utilities.getCallTypeImage(recentCall.getCallType()));
        holder.showHeader(false);

        if (mOnRecentItemClickListener != null) {
            holder.setOnItemClickListener(new ListItemHolder.OnItemClickListener() {
                @Override
                public void onItemClickListener() {
                    mOnRecentItemClickListener.onRecentItemClick(recentCall);
                }

                @Override
                public void onItemLongClickListener() {
                    mOnRecentItemClickListener.onRecentItemLongClick(recentCall);
                }
            });
        }
    }

    @Override
    public int getIdColumn() {
        return mCursor != null ? mCursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.CONTACT_ID) : 1;
    }

    public void setOnRecentItemClickListener(OnRecentItemClickListener onRecentItemClickListener) {
        mOnRecentItemClickListener = onRecentItemClickListener;
    }

    public interface OnRecentItemClickListener {
        void onRecentItemClick(RecentCall recentCall);

        void onRecentItemLongClick(RecentCall recentCall);
    }

}

File: app/src/main/java/com/chooloo/www/callmanager/adapter/RecentsAdapter.kt
package com.chooloo.www.callmanager.adapter

import android.content.Context
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.chooloo.www.callmanager.entity.Recent
import com.chooloo.www.callmanager.ui.listitem.ListItem
import com.chooloo.www.callmanager.ui.listitem.ListItemHolder
import com.chooloo.www.callmanager.util.AnimationUtils.setFadeUpAnimation
import com.chooloo.www.callmanager.util.ContactUtils.lookupContact
import com.chooloo.www.callmanager.util.RelativeTime.getTimeAgo
import com.chooloo.www.callmanager.util.Utilities

class RecentsAdapter(
        private val context: Context
) : RecyclerView.Adapter<ListItemHolder>() {

    private var _recents: Array<Recent> = arrayOf()

    private var _onRecentItemClickListener: ((Recent) -> Unit?)? = null
    private var _onRecentItemLongClickListener: ((Recent) -> Unit?)? = null

    override fun getItemCount(): Int {
        return _recents.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListItemHolder {
        return ListItemHolder(ListItem(parent.context))
    }

    override fun onBindViewHolder(holder: ListItemHolder, position: Int) {
        val recentCall = _recents.get(position)
        val contact = lookupContact(context, recentCall.callerNumber)

        holder.listItem.run {
            setBigText(contact.name ?: recentCall.callerNumber)
            setSmallText(if (recentCall.callDate != null) getTimeAgo(recentCall.callDate.getTime()) else null)
            setImageDrawable(ContextCompat.getDrawable(context, Utilities.getCallTypeImage(recentCall.callType)))
            setOnClickListener { _onRecentItemClickListener?.invoke(recentCall) }
            setOnLongClickListener {
                _onRecentItemLongClickListener?.invoke(recentCall)
                true
            }
            setFadeUpAnimation(this)
        }
    }

    fun updateRecents(newRecents: Array<Recent>) {
        _recents = newRecents
        notifyDataSetChanged()
    }

    fun setOnRecentItemClickListener(onRecentItemClickListener: ((Recent) -> Unit?)?) {
        _onRecentItemClickListener = onRecentItemClickListener
    }

    fun setOnRecentItemLongClickListener(onRecentItemLongClickListener: ((Recent) -> Unit?)?) {
        _onRecentItemLongClickListener = onRecentItemLongClickListener
    }
}